File: 1/2020.10.23/152.Maximum Product Subarray/python/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_p = min_p = res = nums[0]
        for num in nums[1:]:
            cur_max = max_p
            max_p = max(num, num * cur_max, num * min_p)
            min_p = min(num, num * cur_max, num * min_p)
            res = max(res, max_p)
        return res

File: 1/2020.10.23/152.Maximum Product Subarray/java/solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxP = nums[0], minP = nums[0], res = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < 0) {
                int temp = maxP;
                maxP = minP;
                minP = temp;
            }
            maxP = Math.max(nums[i], nums[i] * maxP);
            minP = Math.min(nums[i], nums[i] * minP);
            res = Math.max(res, maxP);
        }
        return res;
    }
}
